### ISSUE TITLE
Support for terraform-plugin-sdk/v2

### ISSUE THREAD

**Issue by xescugc** [terraform-providers/terraform-provider-azurerm#10022]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It's been a while since the v2 of the terraform-plugin-sdk has been released (current one is 2.4.0) and I was wondering if `azurerm` was planning on migrating to it ([docs](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html)).

I don't know if this is easy/hard or a big breaking change but I think it should be done in order to not drag this too much.

**Comment by ross-p-smith** [comment]
I just came across this and looked at what it entailed to upgrade - I noticed that we will need https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#removal-of-helper-hashcode-package as a first step.

**Comment by xescugc** [comment]
Yes (I was working on all the fixes) It may also requires an update on AzureAD Provider as it includes it as a dependency, but not sure 100% as I did not finish the PR

**Comment by tombuildsstuff** [comment]
hey @xescugc 

Thanks for opening this issue.

Just a heads up that we don't support vendoring the providers and using them in this manner, so in this instance you'd be better off here using the GRPC layer to interact with the provider, rather than vendoring it in.

Moving to Plugin SDKv2 is unfortunately more involved for the larger providers, particularly with regards to testing which is the major blocker at this point in time. We're working through that as we can (and we appreciate the contributions in this direction) - but upgrading to Plugin SDKv2 is a "nice to have", as upgrading doesn't directly benefit users - so it's not something we're directly prioritising at this time and so unfortunately have no timeframe for this.

As such whilst I'd like to thank you for opening this issue - since this issue was opened to support integrating with the provider in an unsupported manner, I'm going to close this issue for the moment, but as mentioned above I'd suggest integrating with the provider via the GPRC interface instead, which is likely more stable long-term.

Thanks!

**Comment by xescugc** [comment]
> Just a heads up that we don't support vendoring the providers and using them in this manner, so in this instance you'd be better off here using the GRPC layer to interact with the provider, rather than vendoring it in.

I know that already as it's a design choice on TF side. But as you can see this repo is doing it too https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/go.mod#L24 hehe.

Nevertheless upgrading to V2 is recommended by TF itself so even if it does not impact the users it's something it should be done.

I'm working on a PR to add it to the Project, most of the changes are interface changes and so on.

At some point they'll drop support for V1 and this will need to be dealt with so better do it now IMO.

If you don't mind I'm working on it and I can open a PR to review afterwards so it can be checked

**Comment by tombuildsstuff** [comment]
@xescugc whilst we appreciate the offer, unfortunately we won't be able to accept that PR due to the sheer number of breaking changes necessary to support Plugin SDKv2 at this time, since there's numerous blockers we're working through to unblock that, as time allows but since it's not directly providing value to users (and requires breaking/conflicting with every open PR due to the breaking changes present, including import-path changes) there's no rush/timeframe to do this at this point.

WRT this provider vendoring the AzureAD provider, that's actually just used in the test framework which has a hook for this manner - this is something we're actively trying to remove but there's several dependencies to doing so.

As discussed above, instead I'd suggest switching to use the GRPC interface instead which is supported in this manner :)

**Comment by xescugc** [comment]
> As discussed above, instead I'd suggest switching to use the GRPC interface instead which is supported in this manner :)

I cannot on my specific usecase but no worries I'll keep a Fork with the V2 as I need it then :)

If at some point you want to migrate I'll potentially have it already done haha.

**Comment by xescugc** [comment]
Also fond:

* https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/media/media_transform_resource.go#L13
* https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/media/media_job_resource.go#L13

A PR fixing those (basically coping it to the internal helpers of `azurerm` as intended) would be accepted? It'll still not remove the dependency to `azuread` as it's used on the tests but at least it would only be used there and not on the code.
